kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/AwsKms.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br/>
+     * @see <a href="https://docs.aws.amazon.com/kms/latest/APIReference/API_DescribeKey.html">https://docs.aws.amazon.com/kms/latest/APIReference/API_DescribeKey.html</a>
+     */
+    @NonNull
+    @Override
+    public CompletableFuture<String> resolveAlias(@NonNull String alias) {
+        final DescribeKeyRequest resolveRequest = new DescribeKeyRequest(ALIAS_PREFIX + alias);
+        var request = createRequest(resolveRequest, TRENT_SERVICE_DESCRIBE_KEY);
+        return sendAsync(alias, request, DESCRIBE_KEY_RESPONSE_TYPE_REF, UnknownAliasException::new)
+                .thenApply(DescribeKeyResponse::keyMetadata)
+                .thenApply(KeyMetadata::keyId);
+    }
+
+    private <T> CompletableFuture<T> sendAsync(@NonNull String key, HttpRequest request,
+                                               TypeReference<T> valueTypeRef,
+                                               Function<String, KmsException> exception) {
+        return client.sendAsync(request, HttpResponse.BodyHandlers.ofByteArray())
+                .thenApply(response -> checkResponseStatus(key, response, exception))
+                .thenApply(HttpResponse::body)
+                .thenApply(bytes -> decodeJson(valueTypeRef, bytes));
+    }
+
+    private static <T> T decodeJson(TypeReference<T> valueTypeRef, byte[] bytes) {
+        try {
+            return OBJECT_MAPPER.readValue(bytes, valueTypeRef);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @NonNull
+    private static HttpResponse<byte[]> checkResponseStatus(@NonNull String key,
+                                                            @NonNull HttpResponse<byte[]> response,
+                                                            @NonNull Function<String, KmsException> notFound) {
+        var statusCode = response.statusCode();
+        // AWS API states that only the 200 response is currently used.
+        // Our HTTP client is configured to follow redirects so 3xx responses are not expected here.
+        var httpSuccess = statusCode >= 200 && statusCode < 300;
+        if (!httpSuccess) {
+            ErrorResponse error;
+            try {
+                error = decodeJson(ERROR_RESPONSE_TYPE_REF, response.body());
+            }
+            catch (UncheckedIOException e) {
+                error = null;
+            }
+
+            if (error != null && error.isNotFound()) {
+                throw notFound.apply("key '%s' is not found (AWS error: %s).".formatted(key, error));
+            }
+
+            throw new KmsException("Operation failed, key %s, HTTP status code %d, AWS error: %s".formatted(key, statusCode, error));
+        }
+        return response;
+    }
+
+    @NonNull
+    @Override
+    public Serde<AwsKmsEdek> edekSerde() {
+        return new AwsKmsEdekSerde();
```

Comment:
nit: almost certainly down in the weeds for memory usage but should we just have a singleton instance of the serde?

Reply:
Yes, good spot.  Vault has the same issue. I'll open a separate PR to address both.